core/src/main/scala/org/apache/spark/rdd/CoalescedRDD.scala

```diff
@@ -273,7 +273,8 @@ private class PartitionCoalescer(maxPartitions: Int, prev: RDD[_], balanceSlack:
       groupArr += pgroup
       groupHash.getOrElseUpdate(nxt_replica, ArrayBuffer()) += pgroup
       var tries = 0
-      while (!addPartToPGroup(nxt_part, pgroup) && tries < targetLen) { // ensure at least one part
+      // ensure at least one part
+      while (pgroup.arr.isEmpty && !addPartToPGroup(nxt_part, pgroup) && tries < targetLen) {
```

Comment:
@srowen It is MutableArrayBuffer, looking at http://www.scala-lang.org/api/2.11.0/index.html#scala.collection.mutable.ArrayBuffer , I think isEmpty should be "safe" since it explicitly inherits from IndexedSeqOptimized which provides a concrete isEmpty without an implicit conversion.